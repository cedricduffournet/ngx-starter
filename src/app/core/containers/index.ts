export * from './toaster';
export * from './http-progress-bar/http-progress-bar.component';
