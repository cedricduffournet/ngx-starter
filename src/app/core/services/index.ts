export * from './http-progress.interceptor';
export * from './http.service';
