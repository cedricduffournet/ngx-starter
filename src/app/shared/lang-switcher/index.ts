export * from './lang-switcher.module';
