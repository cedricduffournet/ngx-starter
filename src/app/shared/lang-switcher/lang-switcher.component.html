<div class="lang-switcher" *ngIf="current$ | async as current">
  <button
    *ngFor="let lang of langs"
    type="button"
    [ngClass]="{ 'btn-light': current !== lang, 'btn-info': current === lang }"
    class="btn btn-sm {{ current }}"
    (click)="onSwitch(lang)"
  >
    {{ lang | uppercase }}
  </button>
</div>
