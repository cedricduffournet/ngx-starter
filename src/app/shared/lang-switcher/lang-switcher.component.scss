@import 'src/variables.scss';

.lang-switcher button:not(:first-child) {
  margin-left:map-get($spacers, 1); 
}