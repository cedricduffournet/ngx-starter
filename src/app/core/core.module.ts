import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { environment } from '@env/environment';
import {
  HttpClientModule,
  HttpClient,
  HTTP_INTERCEPTORS
} from '@angular/common/http';

import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { ModalModule } from 'ngx-bootstrap/modal';
import { ToasterModule } from 'angular2-toaster';

import {
  ToasterComponent,
  HttpProgressBarComponent
} from '@app/core/containers';
import { HttpProgressInterceptor } from '@app/core/services/http-progress.interceptor';

const MODULES: any[] = [
  HttpClientModule,
  TranslateModule.forRoot({
    loader: {
      provide: TranslateLoader,
      useFactory: HttpLoaderFactory,
      deps: [HttpClient]
    }
  }),
  ModalModule.forRoot(),
  ToasterModule.forRoot(),
  CommonModule
];

export const COMPONENTS = [ToasterComponent, HttpProgressBarComponent];

@NgModule({
  imports: MODULES,
  declarations: COMPONENTS,
  exports: COMPONENTS,
  providers: [
    {
      provide: HTTP_INTERCEPTORS,
      useClass: HttpProgressInterceptor,
      multi: true
    }
  ]
})
export class CoreModule {
  constructor(
    @Optional()
    @SkipSelf()
    parentModule: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        'CoreModule is already loaded. Import only once in main AppModule.'
      );
    }
  }
}

export function HttpLoaderFactory(http: HttpClient) {
  return new TranslateHttpLoader(
    http,
    `${environment.prefixAsset}/assets/i18n/`,
    '.json'
  );
}
