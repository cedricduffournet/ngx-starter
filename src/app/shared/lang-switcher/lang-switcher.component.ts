import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';

import { BehaviorSubject } from 'rxjs';

import { TranslateService, LangChangeEvent } from '@ngx-translate/core';
import { environment } from '@env/environment';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'app-lang-switcher',
  templateUrl: './lang-switcher.component.html',
  styleUrls: ['lang-switcher.component.scss']
})
export class LangSwitcherComponent implements OnInit {
  current$: BehaviorSubject<string>;
  langs = environment.lang;

  constructor(public translate: TranslateService) {}

  onSwitch(lang: string) {
    this.translate.use(lang);
  }

  ngOnInit() {
    this.current$ = new BehaviorSubject<string>(this.translate.currentLang);
    this.translate.onLangChange.subscribe((evt: LangChangeEvent) =>
      this.current$.next(evt.lang)
    );
  }
}
