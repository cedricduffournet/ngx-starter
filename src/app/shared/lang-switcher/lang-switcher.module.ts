import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { LangSwitcherComponent } from '@app/shared/lang-switcher/lang-switcher.component';

@NgModule({
  imports: [CommonModule],
  exports: [LangSwitcherComponent],
  declarations: [LangSwitcherComponent]
})
export class LangSwitcherModule {}
