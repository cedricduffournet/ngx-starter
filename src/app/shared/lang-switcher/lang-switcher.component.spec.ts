import { TestBed, ComponentFixture, async } from '@angular/core/testing';

import { TranslateModule } from '@ngx-translate/core';

import { LangSwitcherComponent } from '@app/shared/lang-switcher/lang-switcher.component';

describe('LangSwitcherComponent', () => {
  let fixture: ComponentFixture<LangSwitcherComponent>;
  let component: LangSwitcherComponent;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [TranslateModule.forRoot()],
      declarations: [LangSwitcherComponent]
    });
    fixture = TestBed.createComponent(LangSwitcherComponent);
    component = fixture.componentInstance;
  });

  it('should be created', () => {
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should be created with en selected', () => {
    component.translate.currentLang = 'en';

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should be created with en selected', () => {
    component.translate.currentLang = 'en';

    fixture.detectChanges();

    const buttonEn = fixture.nativeElement.querySelector('.en.btn-info');
    expect(buttonEn).toBeTruthy();
  });

  it('should change lang on onSwitch', async(() => {
    fixture.detectChanges();
    component.onSwitch('fr');
    component.current$.subscribe(value => {
      expect(value).toBe('fr');
    });

    fixture.detectChanges();
    const buttonFr = fixture.nativeElement.querySelector('.fr.btn-info');
    expect(buttonFr).toBeTruthy();
  }));
});
