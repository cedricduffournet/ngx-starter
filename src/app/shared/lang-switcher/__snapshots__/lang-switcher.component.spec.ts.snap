// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LangSwitcherComponent should be created 1`] = `
<app-lang-switcher
  current$={[Function BehaviorSubject]}
  langs={[Function Array]}
  translate={[Function TranslateService]}
>
  
</app-lang-switcher>
`;

exports[`LangSwitcherComponent should be created with en selected 1`] = `
<app-lang-switcher
  current$={[Function BehaviorSubject]}
  langs={[Function Array]}
  translate={[Function TranslateService]}
>
  <div
    class="lang-switcher"
  >
    
    <button
      class="btn btn-sm en btn-light"
      ng-reflect-klass="btn btn-sm en"
      ng-reflect-ng-class="[object Object]"
      type="button"
    >
       FR 
    </button>
    <button
      class="btn btn-sm en btn-info"
      ng-reflect-klass="btn btn-sm en"
      ng-reflect-ng-class="[object Object]"
      type="button"
    >
       EN 
    </button>
  </div>
</app-lang-switcher>
`;
